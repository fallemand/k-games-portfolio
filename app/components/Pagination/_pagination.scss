@import '../../styles/variables';

.pagination {
  display: inline-block;

  .pagination__number {
    color: $font-color-primary;
    font-weight: $font-light;
    float: left;
    font-size: 13px;
    padding: 8px 16px;
    text-decoration: none;
    transition: background-color .3s;
    border: 1px solid $background-color-tertiary;
    margin: 0 4px;
    cursor: pointer;

    &.pagination__number--active {
      background-color: $background-color-primary;
      color: white;
      border: 1px solid $background-color-primary;

      &:focus {
        outline: none;
      }
    }

    &:hover:not(.pagination__number--active) {
      background-color: $background-color-tertiary;
    }
  }
}
