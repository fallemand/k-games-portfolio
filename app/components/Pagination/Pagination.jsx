import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import './_pagination.scss';

const Pagination = ({ className, total, show, onChange, active }) => {
  const pages = Math.ceil(total / show);
  return (
    <div className={classnames(className, 'pagination')}>
      <button
        type="button"
        className="pagination__number"
        onClick={() => onChange(active > 1 ? active - 1 : 1)}
      >
        &laquo;
      </button>
      {Array.from(Array(pages), (x, i) => i + 1).map(
        pageNumber => (
          <button
            type="button"
            className={classnames(
              'pagination__number',
              { 'pagination__number--active': (active === pageNumber) },
            )}
            onClick={() => onChange(pageNumber)}
            key={pageNumber}
          >
            {pageNumber}
          </button>),
      )}
      <button
        type="button"
        className="pagination__number"
        onClick={() => onChange(active < pages ? active + 1 : pages)}
      >
        &raquo;
      </button>
    </div>
  );
};

Pagination.propTypes = {
  className: PropTypes.string,
  total: PropTypes.number.isRequired,
  show: PropTypes.number.isRequired,
  active: PropTypes.number,
  onChange: PropTypes.func.isRequired,
};

Pagination.defaultProps = {
  className: null,
  active: 1,
};

export default Pagination;
