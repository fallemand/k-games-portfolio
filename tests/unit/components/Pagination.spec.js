import React from 'react';
import Pagination from '../../../app/components/Pagination';

describe('Pagination', () => {
  const props = {
    className: '__CLASSNAME__',
    total: 20,
    show: 2,
    onChange: jest.fn(),
    active: 5,
  };

  it('render with props', () => {
    const component = shallow(<Pagination {...props} />);
    expect(escapeSnapshot(component)).toMatchSnapshot();
  });

  it('check pages generated if active page is in the middle (2 pages before, 2 after)', () => {
    const component = shallow(<Pagination {...props} />);
    const buttons = component.find('.pagination__number');
    expect(buttons).toHaveLength(7);
    expect(buttons.map(button => button.text())).toEqual(['«', '3', '4', '5', '6', '7', '»']);
  });

  it('check pages generated if active page is 1 (only 2 after active page)', () => {
    const newProps = Object.assign({}, props, { active: 1 });
    const component = shallow(<Pagination {...newProps} />);
    const buttons = component.find('.pagination__number');
    expect(buttons).toHaveLength(5);
    expect(buttons.map(button => button.text())).toEqual(['«', '1', '2', '3', '»']);
  });

  it('check pages generated if active page is the last (only 2 before)', () => {
    const lastPage = props.total / props.show; // 10
    const newProps = Object.assign({}, props, { active: lastPage });
    const component = shallow(<Pagination {...newProps} />);
    const buttons = component.find('.pagination__number');
    expect(buttons).toHaveLength(5);
    expect(buttons.map(button => button.text())).toEqual(['«', '8', '9', '10', '»']);
  });

  it('click on another page should call onChange with page number', () => {
    const component = shallow(<Pagination {...props} />);
    // Active page is 5, so next number is 6.
    const pageTwoNumber = component.find('.pagination__number--active + .pagination__number');
    pageTwoNumber.simulate('click');
    expect(props.onChange).toHaveBeenCalledWith(6);
  });
});
