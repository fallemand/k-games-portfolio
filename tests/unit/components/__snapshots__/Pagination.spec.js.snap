// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination render with props 1`] = `
"<div className='__CLASSNAME__ pagination'>
  <button type='button' disabled={false} className='pagination__number' onClick={[Function: onClick]}>
    «
  </button>
  <button type='button' className='pagination__number' onClick={[Function: onClick]}>
    3
  </button>
  <button type='button' className='pagination__number' onClick={[Function: onClick]}>
    4
  </button>
  <button type='button' className='pagination__number pagination__number--active' onClick={[Function: onClick]}>
    5
  </button>
  <button type='button' className='pagination__number' onClick={[Function: onClick]}>
    6
  </button>
  <button type='button' className='pagination__number' onClick={[Function: onClick]}>
    7
  </button>
  <button type='button' className='pagination__number' disabled={false} onClick={[Function: onClick]}>
    »
  </button>
</div>"
`;
